### PR TITLE
commands/operator-sdk: rename to cmd/operator-sdk and reorganize

### LINKED ISSUE #1184: Do not expose commands/operator-sdk/cmd
**Problem**
Users are not going to need `commands/operator-sdk/cmd` as a library, so it should not be exported.

**Proposed fix**
Move `commands/operator-sdk/cmd` -> `commands/operator-sdk/internal/cmd`

### PR DESCRIPTION
**Pull request by joelanford**
**Description of the change:**
- Moving `commands/operator-sdk/cmd` to `cmd/operator-sdk/`
- Collapsing `commands/operator-sdk/cmd/<subcommand>` to `commands/operator-sdk/<subcommand>`

**Motivation for the change:**
Necessary for 1.0.0 release and closes #1184

### CODE REVIEW

File: commands/operator-sdk/internal/cmd/generate/openapi.go [hunk 1]
```diff
@@ -22,7 +22,7 @@ import (
 	"path/filepath"
 	"strings"
 
-	genutil "github.com/operator-framework/operator-sdk/commands/operator-sdk/cmd/generate/internal"
+	genutil "github.com/operator-framework/operator-sdk/commands/operator-sdk/internal/cmd/generate/internal"
```
Comment (AlexNPavel) [inline commands/operator-sdk/internal/cmd/generate/openapi.go]: We can probably move `genutil.go` from `generate/internal` to `generate` since now the entire `commands/operator-sdk/cmd` is internal.

/cc @estroz

Comment (joelanford) [reply commands/operator-sdk/internal/cmd/generate/openapi.go]: @AlexNPavel I don't think it hurts anything to have nested `internal` directory trees. If we leave `generate/internal` as is, it means that those functions cannot be used outside of the `generate` package. I don't have a strong opinion either way though.

@estroz Thoughts? Was your intent to hide those functions from end users or from other parts of the SDK?

Comment (LiliC) [reply commands/operator-sdk/internal/cmd/generate/openapi.go]: Seems like the functions are all used in the `generate` package so I agree moving them to the `internal/cmd/generate/genutil.go` makes sense, but I would also make them private, as at least from a quick search I can't seem to find them being used anywhere else but in that package.

**Comment by shawn-hurley** [comment]
This feels wrong to me. 

I don't think that anyone expects the cmd/ package to be importable, and no other CLI/Library does this that I can tell. I don't think this will break anything, but it just seems unnecessary and will confuse folks who come to contribute IMO.

**Comment by joelanford** [comment]
@shawn-hurley the [Go project layout convention](https://github.com/golang-standards/project-layout#go-directories) states:
> Don't put a lot of code in the application directory. If you think the code can be imported and used in other projects, then it should live in the /pkg directory. If the code is not reusable or if you don't want others to reuse it, put that code in the /internal directory. You'll be surprised what others will do, so be explicit about your intentions!

I agree with that statement, especially that last sentence. To better align with this, I'll move `commands/operator-sdk/cmd` under the root `/internal` directory, as described, and we'll see how that looks.

**Comment by shawn-hurley** [comment]
I just don't see why many packages don't need to do this and we decided we do. 

https://github.com/kubernetes-sigs/kubebuilder
both pkg/ and cmd/

https://github.com/kubernetes/kubernetes
both pkg/ and cmd/

https://github.com/kubernetes-incubator/service-catalog
both pkg/ and cmd/

and I am sure we can find more. We should look like the rest of the ecosystem.

**Comment by joelanford** [comment]
@shawn-hurley Maybe we don't need to do this, but it seems like a harmless change that has the added benefit of explicitly removing packages from our public API that we don't intend for others to use. I think by leaving things as they are, we are making the statement that they're importable (humans may or may not know our intentions, but tools like `goimports` definitely don't).

There are [lots of imports of `k8s.io/kubernetes/cmd`](https://github.com/search?l=Go&q=%22k8s.io%2Fkubernetes%2Fcmd%22+-repo%3Akubernetes%2Fkubernetes&type=Code) in other repos, so in that case it seems the Kubernetes maintainers have either decided to make those packages part of their public API or they have been ineffective in getting users not to import them.

In any case, I'm okay closing this PR if the consensus is that we're okay leaving the existing layout in place.

**Comment by hasbro17** [comment]
We technically only have 1 command/executable(pkg with main identifier) that's `commands/operator-sdk/main.go`.
All other pkgs that we have at `commands/operator-sdk/cmd/*` are just utils/helpers for that main cmd. And those are not meant to be imported outside the project.

Ideally we should rename `commands/operator-sdk/main.go`=>`cmd/operator-sdk/main.go`.
And the non-main helper pkgs then become `cmd/operator-sdk/add`, `cmd/operator-sdk/generate` etc. That format is more in line with the [unofficial guidelines](https://github.com/golang-standards/project-layout/blob/master/cmd/README.md) for cmd pkgs.


And since those helper pkgs are meant to only be used by `cmd/operator-sdk` ideally they'd become `cmd/operator-sdk/internal/generate`.
But in practice I can see shawn's point on the rest of the ecosystem not doing this.
https://github.com/kubernetes/kubernetes/tree/master/cmd/cloud-controller-manager/app
Beyond the kube ecosystem I can even see some examples in upstream Go code where you have exported identifiers in cmd pkgs.
https://github.com/golang/tools/blob/master/cmd/benchcmp/compare.go#L21 

So I'm okay with relaxing the requirement of having `cmd/operator-sdk/internal/*` and instead relying on the expectation that those pkgs aren't meant to be imported.

Although I'd want to find out if the convention in the kube ecosystem is just an oversight or if those identifiers are meant to be for public use.

**Comment by LiliC** [comment]
My two cents:

I agree about following the guidlines [here](https://github.com/golang-standards/project-layout/blob/master/cmd/README.md#cmd), it seems to be well accepted in the community. It does also mention the use of `internal` directory in `cmd` (which I agree, we should rename `command`->`cmd`, as we do in the end generate an operator with `cmd` dir name, so we should be consistent besides following conventions)

> ...If the code is not reusable or if you don't want others to reuse it, put that code in the /internal directory. You'll be surprised what others will do, so be explicit about your intentions!

Also the discussion on this [issue](https://github.com/golang-standards/project-layout/issues/10) is interesting, I think Jaana (rakyll) made a good point and I feel like it's how majority of users see/use packages as well (regardless where the package might live):

>  Any importable package is for public use. Anything else goes into internal. Note that internal package is not just a convention, it is supported by the Go tool.


To sum up, if we are already making changes to the API and making things private/internal towards the `1.0.0` release we might as well cleanup a bit and follow the conventions. Personally I agree with the above and we should do the following in this PR instead:
```
commands/operator-sdk/cmd/add.go → cmd/operator-sdk/add/...
```
and any internal components we can move to:
```
cmd/operator-sdk/internal/...
```

**Comment by joelanford** [comment]
Ok I've updated the PR to account for the recent comments. The `commands` directory has been renamed to `cmd`, and the subcommand packages have been collapsed down to `cmd/operator-sdk/<subcommand>`.

I moved the `New*Cmd` functions into the respective subcommand packages and (where possible) made the rest of the functions and idenifiers private. There were two exceptions:
1. In `cmd/operator-sdk/generate`, there are two functions (`K8sCodegen()` and `OpenAPIGen()`) that are called by the `add` subcommand, so they needed to remain public. Thoughts on leaving that dependency in place or extracting those to a shared internal library?
2. In `cmd/operator-sdk/scorecard`, there's a large number of exported identifiers. I imagine we'll want to move the public-facing helpers and APIs to `pkg` at some point, and unexport the rest. But since @AlexNPavel is making lots of changes here, I figured light touch was best.

@shawn-hurley @hasbro17 @LiliC @AlexNPavel PTAL.

**Comment by LiliC** [comment]
Looks good! 🎉 

> In cmd/operator-sdk/generate, there are two functions (K8sCodegen() and OpenAPIGen()) that are called by the add subcommand, so they needed to remain public. Thoughts on leaving that dependency in place or extracting those to a shared internal library?

Yes I would extract them to a shared internal lib, think `cmd/operator-sdk/internal` should work and it can be used by both the `generate` and the `add` packages. It shouldn't be part of our API I think. WDYT?

**Comment by joelanford** [comment]
@AlexNPavel @LiliC @shawn-hurley @hasbro17 PTAL

I extracted the generator helpers into `/cmd/operator-sdk/internal/genutil`. 

As a result, I had to refactor the way that the header file flag was handled in the OpenAPI generator. Now that the generator is in a separate package from its callers, it didn't make sense for the caller and callee to share a global variable for the header file. 

Instead, I made an options struct that includes the header file and updated the `genutil.OpenAPIGen()` function to take that as a parameter. I figured that interface made sense since we're wrapping the `openapi-gen` CLI and may want to add/remove/change what flags we pass through from our CLI. Thoughts?

File: cmd/operator-sdk/internal/genutil/k8s.go [hunk 2]
```diff
@@ -21,15 +21,14 @@ import (
...
-	genutil "github.com/operator-framework/operator-sdk/commands/operator-sdk/cmd/generate/internal"
 	"github.com/operator-framework/operator-sdk/internal/util/projutil"
 	"github.com/operator-framework/operator-sdk/pkg/scaffold"
 
 	log "github.com/sirupsen/logrus"
 	"github.com/spf13/cobra"
 )
 
-func NewGenerateK8SCmd() *cobra.Command {
+func newGenerateK8SCmd() *cobra.Command {
```
Comment (hasbro17) [inline cmd/operator-sdk/internal/genutil/k8s.go]: We have the functions `newGenerateK8SCmd()` and `k8sFunc()`(line 52) defined here in `cmd/operator-sdk/internal/genutil/k8s.go` and in `cmd/operator-sdk/generate/k8s.go`. I think they're only being used in the latter pkg. Do we need this here?

Comment (joelanford) [reply cmd/operator-sdk/internal/genutil/k8s.go]: We don't. Those were mistakenly copied over. Good catch.

**Comment by estroz** [comment]
This build error is unrelated to these changes. Seems like an upstream breaking change, since [my `go mod` PR](https://travis-ci.org/operator-framework/operator-sdk/jobs/511579705#L905) which was previously passing is now failing for the same reason.

**Comment by joelanford** [comment]
> This build error is unrelated to these changes. Seems like an upstream breaking change, since [my `go mod` PR](https://travis-ci.org/operator-framework/operator-sdk/jobs/511579705#L905) which was previously passing is now failing for the same reason.

Here's the error messages: https://travis-ci.org/operator-framework/operator-sdk/jobs/511514286#L910-L913

I'm fairly sure it's related to the [changes we made to support all the client auth plugins](https://github.com/operator-framework/operator-sdk/pull/1199/files#diff-836546cc53507f6b2d581088903b1785).

I'll take a look at what's going on.

File: cmd/operator-sdk/main.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
+// Copyright 2018 The Operator-SDK Authors
```
Comment (hasbro17) [inline cmd/operator-sdk/main.go]: Update license headers to 2019 for new files.

**Comment by estroz** [comment]
LGTM after rebase.
